kubernetes networking model clarifications (#28454)
kubernetes/website: content/en/docs/concepts/cluster-administration/networking.md

## The Kubernetes network model

Every `Pod` gets its own IP address, maximum one per IP family. This means
you do not need to explicitly create links between `Pods` and you almost
never need to deal with mapping container ports to host ports.  This creates
a clean, backwards-compatible model where `Pods` can be treated much like
VMs or physical hosts from the perspectives of port allocation, naming,
service discovery, load balancing, application configuration, and migration.

Kubernetes IP addresses exist at the `Pod` scope - containers within a `Pod`
share their network namespaces - including their IP address. This means that
containers within a `Pod` can all reach each other's ports on `localhost`.
This also means that containers within a `Pod` must coordinate port usage,
but this is no different from processes in a VM.  This is called the
"IP-per-pod" model.

In every cluster, there exists an abstract pod-network to which pods
are connected by default, unless explicitly configured to use the
host-network (on platforms that support it). Even if a host has

> multiple IPs, host-network pods only have one Kubernetes IP address at

Reviewer: Perhaps a cleaner wording:

> Even if a host has multiple IPs, the Pod's Status.PodIPs field will only ever contain one IP per address family.

Author: there are not references to the fields before, that is why omitted that approach ... let's see what the other say, 

What  should be clear is that `Kubernetes IPs = Pod.Status.PodIPs`

Reviewer 2: If we do mention the API, let's capitalize fields the way they appear in the API (`status.podIPs`).

Reviewer: I personally prefer making explicit references to the API. As we all know, if it's in etcd, it must be true :-).

More seriously, the API is how people interact with the cluster. In other words, what does it mean for a pod to "have" an IP?

the `Pod` scope, so the "IP-per-pod" model is guaranteed.

Kubernetes imposes the following fundamental requirements on any networking
implementation (barring any intentional network segmentation policies):

   * any pod-network pod on any node can communicate with all other pod-network
     pods on all nodes without NAT.
   * non-pod processes on a node, for example, system daemons, kubelet, ... can
     communicate with all pods on that node.

In addition, for platforms and runtimes that support `Pods` running in the host
network:

>    * host-network pods of a node can connect directly with all pods IPs on all

Reviewer: suggested wording:

> host-network pods, as well as non-pod processes, can connect directly to all pod IPs on all nodes - including other host-network pods. However, the source IP address may not be included in `Pod.Status.PodIPs` or `Node.Status.Addresses`.

## The Kubernetes network model

Every `Pod` gets its own IP address, maximum one per IP family. This means

> you do not need to explicitly create links between `Pods` and you almost

Author: _This means you do not need need to deal with mapping container ports to host ports in order to
expose the `Pods` services on the network_

never need to deal with mapping container ports to host ports.  This creates
a clean, backwards-compatible model where `Pods` can be treated much like
VMs or physical hosts from the perspectives of port allocation, naming,
service discovery, load balancing, application configuration, and migration.

Kubernetes IP addresses exist at the `Pod` scope, in the `status.PodIPs` field
- containers within a `Pod` share their network namespaces - including their IP
address. This means that containers within a `Pod` can all reach each other's
ports on `localhost`. This also means that containers within a `Pod` must
coordinate port usage, but this is no different from processes in a VM. This is

> called the "IP-per-pod" model.

Reviewer 2: nit:
```suggestion
called the _IP-per-pod_ model.
```

In every cluster, there exists an abstract pod-network to which pods
are connected by default, unless explicitly configured to use the
host-network (on platforms that support it). Even if a host has
multiple IPs, host-network pods only have one Kubernetes IP address at

> the `Pod` scope, that is, the `status.PodIPs` field will only ever contain one

Reviewer 3: Can we avoid wording like "will only ever" - that's a promise I don't think we should make.  Maybe "the `status.PodIPs` field only contains one IP per address family (for now)"

IP per address family, so the "IP-per-pod" model is guaranteed.

Kubernetes imposes the following fundamental requirements on any networking
implementation (barring any intentional network segmentation policies):

   * any pod-network pod on any node can communicate with all other pod-network
     pods on all nodes without NAT.
   * non-pod processes on a node, for example, system daemons, kubelet, ... can
     communicate with all pods on that node.

In addition, for platforms and runtimes that support `Pods` running in the host
network:

   * host-network pods of a node can connect directly with all pods IPs on all
     nodes, however, unlike pod-network pods, the source IP address may not be
     present in the `Pod` `status.PodIPs` field.

This model is principally compatible with the desire for Kubernetes to enable

> low-friction porting of apps from VMs to containers. If your job previously ran
> in a VM, your VM had an IP and could talk to other VMs in your project.

Reviewer 2: How about:
```suggestion
low-friction porting of apps from VMs to containers. If your workload previously ran
in a VM, your VM typically had a single IP address; everything in that VM could talk to
other VMs on your network.
```
(writing “workload” rather than “job” because Kubernetes uses that term, “job”, to mean a specific API)

service discovery, load balancing, application configuration, and migration.

Kubernetes IP addresses exist at the `Pod` scope, in the `status.PodIPs` field
- containers within a `Pod` share their network namespaces - including their IP
address. This means that containers within a `Pod` can all reach each other's
ports on `localhost`. This also means that containers within a `Pod` must
coordinate port usage, but this is no different from processes in a VM. This is
called the _IP-per-pod_ model.

In every cluster, there exists an abstract pod-network to which pods
are connected by default, unless explicitly configured to use the
host-network (on platforms that support it). Even if a host has
multiple IPs, host-network pods only have one Kubernetes IP address at
the `Pod` scope, that is, the `status.PodIPs` field will only ever contain one
IP per address family, so the "IP-per-pod" model is guaranteed.

Kubernetes imposes the following fundamental requirements on any networking
implementation (barring any intentional network segmentation policies):

   * any pod-network pod on any node can communicate with all other pod-network
     pods on all nodes without NAT.

>    * non-pod processes on a node, for example, system daemons, kubelet, ... can

Reviewer 2: ```suggestion
   * non-pod processes on a node (the kubelet, and also for example any other system daemon) can
```

>      communicate with all pods on that node.

Reviewer 4: I'd commented on this before but it got eaten by a re-push; as I understand it, this is incorrect for Windows.

The only thing we _require_ is that _kubelet_ can communicate with all pods on the node. The fact that non-pod processes can communicate with the pod network on Linux is a side effect of the requirement that hostNetwork pods can communicate with the pod network, _not_ a side effect of the requirement that kubelet can communicate with pods.

Author: I've resolved it based on this comment from the windows folks https://github.com/kubernetes/website/pull/28454#discussion_r687053241

> if it is overlay, any systems services can only reach pods on the same node. If it is l2bridge then system services can communicate to pods across nodes.

did I misunderstand the comment?

Reviewer 5: There is nothing special about kubelet on Windows that enables it to communicate with pods on the same node in Overlay mode.  On Windows in Overlay mode, Any service on the node would be able to communicate to pods on the same node but not pods on other nodes. cc: @daschott

Reviewer 4: ah, somehow I missed that response.
ok, it's the distinction between "all pods on this node" and "all pods on all nodes" that I'd missed

> In addition, for platforms and runtimes that support `Pods` running in the host

Reviewer 2: (nit) `Pods` (plural but in monospace) is a bit odd, so I recommend fixing that plus a wording tweak:
```suggestion
In addition, for platforms and runtimes that support running pods in the host OS' network
```

Author: OS' , is the last apostrophe meaning something?

Reviewer 2: It's a possessive apostrophe; “the network of the host operating system”

Author: free English class :smile: , I thought it was with an s `OS's` , is this because the last letter is an `s` too?

Reviewer 2: OS's might be fine too. I don't speak American English very well myself (I'm from England :wink:)

Author: I'm from Spain, whatever you've decide :shrug:

> In addition, for platforms and runtimes that support running pods in the host OS' network
> network:

Reviewer 6: ```suggestion
In addition, for platforms and runtimes that support running pods in the host OS network:
```

>    * host-network pods of a node can connect directly with all pods IPs on all

Reviewer 5: now that we have Windows HostProcess Containers this statement might not be correct for Windows when running in overlay (for same reasons as https://github.com/kubernetes/website/pull/28454/files#r688944775) 

I just tried it and was not able to go from hostprocess container to a pod when in overlay.  cc: @marosset @daschott

Reviewer 7: We should probably update https://github.com/kubernetes/website/blob/f4462936596941a5eab5b3ba322a93e2d96437a3/content/en/docs/setup/production-environment/windows/intro-windows-in-kubernetes.md?plain=1#L542 while we are at it

Author: think on this as the minimum you want to enforce the ecosystem to comply with, not as the current status, i.e., _ Least Common Multiple _

Reviewer 3: Is this confirmed now?

Author: @marosset  I will keep this PR focus on the networking model if you don't mind, the windows docs can be updated in a follow-up.

@jsturtevant do you mean that host-network pods on windows are not able to reach other pods in other nodes? or ir just an implementation detail of the cni plugin for windows with overlay?

Right now a lot of distros and installers deploy apiservers as hostnetwork pods, if those pods are not reachable from the rest of the pods , in-cluster config will not work, I think that we should not remove that restriction from linux, but we can relax it for windows ???

Reviewer 2: In case it helps, we don't support running the control plane on Windows / we don't document that we support that.

Author: it was an example, my point is that some operators, charts, manifests assume this, if this doesn't work there has to be one version of those operators, manifest, ... per platform ...  IMHO very bad UX if we drift on something like this, we end with 2 kubernetes flavors .
I think that is important to know if this is a technical limitation on windows or an implementation detail that can be reasonable solved  .... it puzzles me that it works with one cni and not with the other

>      nodes, however, unlike pod-network pods, the source IP address may not be

Reviewer 6: ```suggestion
     nodes, however, unlike pod-network pods, the source IP address might not be
```

perspectives of port allocation, naming, service discovery, load balancing,
application configuration, and migration.

Kubernetes IP addresses exist at the `Pod` scope - containers within a `Pod`

> share their network namespaces - including their IP address and MAC address.

Author: pods use veth interfaces that are ethernet interfaces, they should always have a mac address ... are there any special cases without macs?

> share their network namespaces - including their IP address and MAC address.

Reviewer 2: Do we insist on using veths or is that just the most popular implementation?

Eg: if I make a network plugin that uses a tun device per pod, does that fail some specification or test?

> share their network namespaces - including their IP address and MAC address.

Author: ah, well, you got me there, I've confirmed with an expert  @squeed  that you are right :smile:

This means that containers within a `Pod` can all reach each other's ports on
`localhost`. This also means that containers within a `Pod` must coordinate port
usage, but this is no different from processes in a VM.  This is called the
"IP-per-pod" model.

In every cluster, there exists an abstract pod-network to which pods
are connected by default, unless explicitly configured to use the

> host-network (which is an optional capability). Even if a host have

Reviewer 4: not clear what "(which is an optional capability)" means. I'm guessing you're talking about the fact that the network plugin might not support it, but given that we're mostly talking about pods here it sounds more like you mean "it's optional for pods to specify that they want host-network", which you already just said and so it sounds weird.

maybe "(on platforms that support it)"?

Reviewer 3: Sure - the point was that the ability to specify host-network is optional, but i like your words beter

Author: done

multiple IPs, host-network pods only have one Kubernetes IP address at
the `Pod` scope, so the "IP-per-pod" model is guaranteed.

Kubernetes imposes the following fundamental requirements on any networking
implementation (barring any intentional network segmentation policies):

>    * pod-network pods on a node can communicate with all pod-network pods on
>      all nodes without NAT

Reviewer 2: How about:
```suggestion
  * Any pod can try to send traffic towards any Pod that is also running with
    `hostNetwork: false`. The network addressing scheme must
    allow this.
    Security / filtering mechanisms such as NetworkPolicy could mean that a particular
    pod-to-pod packet is dropped or rejected.
```
This ties in with the rewritten note about SNAT possibly being needed for hostNetwork Pods.

Author: despite the correctness of the suggestion, I found the previous sentence easier to understand.
I've also found it in multiple slides and presentation. 
I think that at this point the `pods on a node can communicate with all pods on all nodes without NAT` is kind of the "1st commandment of the networking model"

Reviewer 3: I see what @sftim is going for, but I think I agree with @aojea.  This is a pretty technical doc and I find the precision and brevity valuable.  

Does it make it better or worse to be more explicit?

* Any pod-network pod on any node can communicate with all other pod-network pods on all nodes without NAT.

Reviewer 2: > Any pod-network pod on any node can communicate with all other pod-network pods on all nodes without NAT.

Sounds good to me. It'd be good to clarify somewhere in the page that being able to address the other Pod doesn't imply that your packets actually arrive there.

Author: is not covered by this sentence?
> Kubernetes imposes the following fundamental requirements on any networking implementation (barring any intentional network segmentation policies):

   * non-pod processes on a node (e.g. system daemons, kubelet) can communicate

>      with all pods on that node

Reviewer 2: nit: we [avoid Latin abbreviations](https://kubernetes.io/docs/contribute/style/style-guide/#avoid-latin-phrases)

Reviewer 4: I'm not sure this is correct for Windows. I had thought that the situation there was that kubelet can communicate with all pods on the node, but no other system processes can reach any pods. I guess we need clarification/confirmation here.

Author: @jsturtevant can you help us here?

Reviewer 8: +1 to @jsturtevant comment.

Author: that is the expected and how this started, the decision was to only require communications from services to pods on the same node, across nodes is not required

> Note: For those platforms that support `Pods` running in the host network (e.g.Linux):

Reviewer 3: another "e.g." :) - we can probably just remove the example.  We all know it means Linux :)

Reviewer 3: Also maybe we should say "platform runtimes" to leave room for runtimes like Kata to choose not to support host-net ?

Author: done

>    * host-network pods of a node can connect directly with all pods IPs

Reviewer 3: Above you left it as "can communicate with" but here changed it to "can connect directly with".  That seems like a very specific change and I'd like to understand why?  My initial feeling is that we should use the same words in both clauses.

Also a nit:  "communicate with" or "connect to".  English :shrug:

Reviewer 4: while we're nitpicking prepositions, "pods of a node" sounds weird to me and I think "pods on a node" would be better...

Author: I don't know how to express direction and origin, 
pod.Status.PodIPs should be reachable ALWAYS, but for host-network pod the origin can be any of the host IP addresses, that is not necessarily published in the API

>    * host-network pods of a node can connect directly with all pods IPs

Reviewer 3: Why is "IPs" important here? In the face of eventual multi-network, I don't think "all pod IPs" is correct.

nit: "all pods" or "all pod IPs" or "all pods' IPs".  English :shrug:

Author: I want to express ,but I don't know how, the difference between a normal IP and  the IPs that belong to the Kubernetes API. Per example, a pod or host can have multiples IPs now, but the kubernetes network model only applies to the ones that are in pod.Status.PodIPs

Author: check this new approach

> Note: For those platforms that support `Pods` running in the host network (e.g.Linux):
> 
>    * host-network pods of a node can connect directly with all pods IPs
>      on all nodes, with source NAT if required

Reviewer 2: Optionally, use a callout. I don't think it's important to mention the OS that the node is using.

Here's what I'd write:
```suggestion
{{< note >}}
Pods running with `hostNetwork: true` can send traffic towards any Pod, even to
Pods running on other nodes. The network plugin is allowed to use techniques
such as source NAT to achieve this. Kubernetes does not specify what the source
IP address of that traffic would be; however, it should not be from the IP address
range reserved for Pods.
{{< /note >}}
```

Author: suggestion make sense, just to clarify that it can be from the range reserved for pods :), actually it   is what happens, since the host use to adopt the .1 of the pod subnet in the node, that is the closest to the pod addresses, so when you create a connection from a hostnetwork pod, the kernel picks that address

Author: also is not only network plugin dependent, there are more factors

Author: the idea is this:
```
host-network pods can communicate with all pods in all nodes, however, the source IP used by the host-network pod can be any IP in the host and use source NAT if required.
```

Reviewer 3: I want to be careful not to treat this as an "aside".  It's an additional hard requirement.

Author: check this approach, I think that NAT is misleading here, the different IPs in a host is the key, NAT is a detail

>      on all nodes, with source NAT if required

Reviewer 4: "with source NAT if required" kind of makes it sound like the pod is responsible for NATting itself

Author: done

> This model is not only less complex overall, but it is principally compatible

Reviewer 4: "less complex" than what? This paragraph seems to be following up on some long-since deleted previous point...

Author: yep, fixed
